pkg/storage/tsdb/bucketcache/caching_bucket.go

```diff
@@ -549,24 +552,86 @@ func mergeRanges(input []rng, limit int64) []rng {
 	return input[:last+1]
 }
 
-func cachingKeyAttributes(name string) string {
-	return fmt.Sprintf("attrs:%s", name)
+func cachingKeyAttributes(bucketID, name string) string {
+	return cachingKeyForSimpleOperation("attrs", bucketID, name)
 }
 
-func cachingKeyObjectSubrange(name string, start, end int64) string {
-	return fmt.Sprintf("subrange:%s:%d:%d", name, start, end)
+func cachingKeyObjectSubrange(bucketID, name string, start, end int64) string {
+	const op = "subrange"
+
+	// Format start and end.
+	startString := strconv.FormatInt(start, 10)
+	endString := strconv.FormatInt(end, 10)
+
+	b := strings.Builder{}
+	b.Grow(len(bucketID) + len(op) + len(name) + len(startString) + len(endString) + 4)
+
+	if bucketID != "" {
+		b.WriteString(bucketID)
+		b.WriteRune(':')
+	}
+
+	b.WriteString(op)
+	b.WriteRune(':')
+	b.WriteString(name)
+	b.WriteRune(':')
+	b.WriteString(startString)
+	b.WriteRune(':')
+	b.WriteString(endString)
+
+	return b.String()
 }
 
-func cachingKeyIter(name string) string {
-	return fmt.Sprintf("iter:%s", name)
+func cachingKeyIter(bucketID, name string, options ...objstore.IterOption) string {
+	const (
+		op        = "iter"
+		recursive = "recursive"
+	)
+
+	b := strings.Builder{}
+	b.Grow(len(bucketID) + len(op) + len(name) + len(recursive) + 3)
+
+	if bucketID != "" {
+		b.WriteString(bucketID)
+		b.WriteRune(':')
+	}
+
+	b.WriteString(op)
+	b.WriteRune(':')
+	b.WriteString(name)
+
+	// Ensure the caching key is different for the same request but different
+	// recursive config.
+	if params := objstore.ApplyIterOptions(options...); params.Recursive {
+		b.WriteRune(':')
+		b.WriteString(recursive)
+	}
+
+	return b.String()
 }
 
-func cachingKeyExists(name string) string {
-	return fmt.Sprintf("exists:%s", name)
+func cachingKeyExists(bucketID, name string) string {
+	return cachingKeyForSimpleOperation("exists", bucketID, name)
 }
 
-func cachingKeyContent(name string) string {
-	return fmt.Sprintf("content:%s", name)
+func cachingKeyContent(bucketID, name string) string {
+	return cachingKeyForSimpleOperation("content", bucketID, name)
+}
+
+func cachingKeyForSimpleOperation(op, bucketID, name string) string {
+	b := strings.Builder{}
+	b.Grow(len(bucketID) + len(op) + len(name) + 2)
+
+	if bucketID != "" {
+		b.WriteString(bucketID)
+		b.WriteRune(':')
+	}
+
+	b.WriteString(op)
+	b.WriteRune(':')
+	b.WriteString(name)
+
+	return b.String()
 }
 
```

Comment:
I admit I much prefer previous version of the code. Adding bucket ID shouldn't be a reason for this code explosion. If there are other reasons, eg. performance, it would be good to see some benchmarks and numbers from real usage. Until then I suggest we revert to "simple stupid" version.

Reply:
Did you see a workload where this code was a hot path contributing to a some problem?

I still don't think this change belong to *this* PR, and I am not convinced that I see the need to optimize this specific code. New version puts extra cognitive load on a reader of this code, and I just don't see the benefit.